packages/ws-worker/test/reasons.test.ts

```diff
@@ -0,0 +1,160 @@
+import test from 'ava';
+import createRTE from '@openfn/engine-multi';
+import createLogger, { createMockLogger } from '@openfn/logger';
+
+import { createPlan } from './util';
+import { execute as doExecute } from '../src/api/execute';
+import { mockChannel } from '../src/mock/sockets';
+
+import {
+  RUN_START,
+  RUN_COMPLETE,
+  ATTEMPT_LOG,
+  ATTEMPT_START,
+  ATTEMPT_COMPLETE,
+} from '../src/events';
+import { ExitReason } from '../src/types';
+
+// Explicit tests of exit reasons coming out of the worker
+// these test the onComplete callback
+// uses the real runtime engine
+
+let engine;
+let logger;
+
+test.before(async () => {
+  logger = createMockLogger();
+  // logger = createLogger(null, { level: 'debug' });
+
+  engine = await createRTE({
+    maxWorkers: 1,
+    purge: false,
+    logger,
+  });
+});
+
+test.after(async () => engine.destroy());
+
+// Wrap up an execute call, capture the on complete state
+const execute = async (plan) =>
+  new Promise<{ reason: ExitReason; state: any }>((done) => {
+    // Ignore all channel events
+    // In these test we assume that the correct messages are sent to the channel
+    const channel = mockChannel({
+      [ATTEMPT_START]: async () => true,
+      [RUN_START]: async () => true,
+      [ATTEMPT_LOG]: async () => true,
+      [RUN_COMPLETE]: async () => true,
+      [ATTEMPT_COMPLETE]: async () => true,
+    });
+
+    const onComplete = (result) => {
+      done(result);
+    };
+
+    // @ts-ignore
+    doExecute(channel, engine, logger, plan, {}, onComplete);
+  });
+
+test('success', async (t) => {
```

Comment:
These tests are a good place to look at the cahnges.

The `reason` object` here is a bucket of keys that are added to the attempt:complete payload

I will add full integrations tests for this but that'll be a later PR